pennylane_lightning_kokkos/src/simulator/MeasuresFunctors.hpp

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <Kokkos_Core.hpp>
+#include <Kokkos_StdAlgorithms.hpp>
+
+#include "Util.hpp"
+
+namespace {
+using namespace Pennylane::Util;
+namespace KE = Kokkos::Experimental;
+} // namespace
+
+namespace Pennylane {
+namespace Functors {
+
+template <class Precision> struct getProbFunctor {
+
+    Kokkos::View<Kokkos::complex<Precision> *> arr;
+    Kokkos::View<Precision *> probilities;
+
+    getProbFunctor(Kokkos::View<Kokkos::complex<Precision> *> arr_,
+                   Kokkos::View<Precision *> probilities_)
+        : arr(arr_), probilities(probilities_) {}
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()(const size_t k) const {
+        Precision REAL = arr[k].real();
+        Precision IMAG = arr[k].imag();
+        probilities[k] = REAL * REAL + IMAG * IMAG;
+    }
+};
+
+template <class Precision> struct getCDFFunctor {
+
+    Kokkos::View<Precision *> probilities;
+
+    getCDFFunctor(Kokkos::View<Precision *> probilities_)
+        : probilities(probilities_) {}
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()(const size_t &k, Precision &update_value,
+                    const bool final) const {
+        const Precision val_k = probilities[k];
+
+        if (final)
+            probilities[k] = update_value;
+
+        update_value += val_k;
+    }
+};
+
+template <class Precision, template <class ExecutionSpace> class GeneratorPool,
+          class ExecutionSpace = Kokkos::DefaultExecutionSpace>
+struct Sampler {
+
+    Kokkos::View<size_t *> samples;
+
+    Kokkos::View<Precision *> cdf;
+
+    GeneratorPool<ExecutionSpace> rand_pool;
+
+    const size_t num_qubits;
+
+    const size_t N;
+
+    Sampler(Kokkos::View<size_t *> samples_, Kokkos::View<Precision *> cdf_,
+            GeneratorPool<ExecutionSpace> rand_pool_, const size_t num_qubits_,
+            const size_t N_)
+        : samples(samples_), cdf(cdf_), rand_pool(rand_pool_),
+          num_qubits(num_qubits_), N(N_) {}
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()(const size_t k) const {
+        // Get a random number state from the pool for the active thread
+        typename GeneratorPool<ExecutionSpace>::generator_type rand_gen =
+            rand_pool.get_state();
```

Comment:
Could the LHS be replaced by `auto` ? Or is it better to keep the explicit type?

Reply:
yes, we can replace it with auto. Let me do it.

Reply:
done